src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs

```diff
@@ -993,6 +993,38 @@ private BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression
             BoundExpression? xNonNull = NullableAlwaysHasValue(loweredLeft);
             BoundExpression? yNonNull = NullableAlwaysHasValue(loweredRight);
 
+            TypeSymbol boolType = _compilation.GetSpecialType(SpecialType.System_Boolean);
+
+            // Optimization: If one side is non-default constant, checking HasValue is not needed.
+            if (kind.Operator() is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual)
```

Comment:
>kind.Operator() is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual [](http://example.com/codeflow?start=16&length=74)

We need to make sure we cover all other binary operators. I do not think we need to cover intersection with all possible argument types, but, perhaps, use distinct type for each operator, if we have to add tests,

Reply:
In the new tests I see tests for `<` and `>=`. I think we should add tests for `<=` and `>` as well.